Error: The terraform-provider-snowflake_ [v0.75.0 - 0.83.1] plugin crashed! on adding New roles and associated grants

### Terraform CLI and Provider Versions

Provider version -> v0.75.0 - v.0.83.1

Terraform CLI version -> 1.2.9

### Terraform Configuration

```terraform
resource "snowflake_role" "warehouse_monitor_role" {
  count   = var.warehouse_config["create_access_roles"] ? 1 : 0
  name    = "${local.warehouse_name}_MONITOR_ONLY"
  comment = "${local.warehouse_name} database role for Monitor Privilege to enable query history"
}

resource "snowflake_grant_privileges_to_role" "warehouse_monitor_role_1" {
  count      = var.warehouse_config["create_access_roles"] ? 1 : 0
  role_name  = snowflake_role.warehouse_monitor_role[count.index].name
  privileges = ["MONITOR"]
  on_account_object {
    object_type = "WAREHOUSE"
    object_name = local.warehouse_name
  }

  depends_on = [
    snowflake_warehouse.warehouse
  ]
}

resource "snowflake_role_grants" "warehouse_monitor_role_2" {
  count     = var.warehouse_config["create_access_roles"] ? 1 : 0
  role_name = snowflake_role.warehouse_monitor_role[count.index].name

  roles = [
    "SYSADMIN"
  ]

  depends_on = [
    snowflake_warehouse.warehouse
  ]
}
```


### Expected Behavior

Update the changes along with add new roles with No error

### Actual Behavior

All plugin versions are generating similar error output:

`Error: -16T14:28:57.293Z [ERROR] plugin6.(*GRPCProvider).ApplyResourceChange: error="rpc error: code = Unavailable desc = transport is closing"
Error: -16T14:28:57.294Z [ERROR] plugin6.(*GRPCProvider).ApplyResourceChange: error="rpc error: code = Unavailable desc = transport is closing"
Error: -16T14:28:57.295Z [ERROR] plugin6.(*GRPCProvider).ApplyResourceChange: error="rpc error: code = Unavailable desc = transport is closing"
Error: -16T14:28:57.295Z [ERROR] plugin6.(*GRPCProvider).ApplyResourceChange: error="rpc error: code = Unavailable desc = transport is closing"
Error: -16T14:28:57.295Z [ERROR] plugin6.(*GRPCProvider).ApplyResourceChange: error="rpc error: code = Unavailable desc = transport is closing"
Error: -16T14:28:57.296Z [ERROR] vertex "module.warehouses[\"DBT_COMMON_WH\"].snowflake_role.warehouse_monitor_role[0]" error: Plugin did not respond
Error: -16T14:28:57.296Z [ERROR] plugin6.(*GRPCProvider).ApplyResourceChange: error="rpc error: code = Unavailable desc = transport is closing"
Error: -16T14:28:57.296Z [ERROR] plugin6.(*GRPCProvider).ApplyResourceChange: error="rpc error: code = Unavailable desc = transport is closing"
Error: -16T14:28:57.296Z [ERROR] plugin6.(*GRPCProvider).ApplyResourceChange: error="rpc error: code = Unavailable desc = transport is closing"
Error: -16T14:28:57.296Z [ERROR] plugin6.(*GRPCProvider).ApplyResourceChange: error="rpc error: code = Unavailable desc = transport is closing"
Error: -16T14:28:57.296Z [ERROR] plugin6.(*GRPCProvider).ApplyResourceChange: error="rpc error: code = Unavailable desc = transport is closing"
Error: -16T14:28:57.298Z [ERROR] vertex "module.warehouses[\"DBT_INVEST_WH\"].snowflake_role.warehouse_monitor_role[0]" error: Plugin did not respond
Error: -16T14:28:57.300Z [ERROR] vertex "snowflake_warehouse.warehouse[\"SCRATCH_WH\"]" error: Plugin did not respond
Error: -16T14:28:57.301Z [ERROR] vertex "snowflake_oauth_integration.dbt_cloud[0]" error: Plugin did not respond
Error: -16T14:28:57.303Z [ERROR] vertex "snowflake_user.user[\"DBT_INVEST\"]" error: Plugin did not respond
Error: -16T14:28:57.304Z [ERROR] vertex "snowflake_warehouse.warehouse[\"DBT_COMMON_WH\"]" error: Plugin did not respond
Error: -16T14:28:57.305Z [ERROR] vertex "snowflake_user.user[\"DBT_COMMON\"]" error: Plugin did not respond
Error: -16T14:28:57.307Z [ERROR] vertex "module.warehouses[\"DATA_INGESTION_WH\"].snowflake_role.warehouse_monitor_role[0]" error: Plugin did not respond
Error: -16T14:28:57.308Z [ERROR] vertex "module.warehouses[\"TS_BI_WH\"].snowflake_role.warehouse_monitor_role[0]" error: Plugin did not respond
Error: -16T14:28:57.309Z [ERROR] vertex "module.warehouses[\"SCRATCH_WH\"].snowflake_role.warehouse_monitor_role[0]" error: Plugin did not respond
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
Releasing state lock. This may take a few moments...

Stack trace from the terraform-provider-snowflake_v0.83.1 plugin:

panic: interface conversion: interface {} is *schema.Set, not []interface {}

goroutine 88 [running]:
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources.UpdateOAuthIntegration(0xc000a28c00, {0x1fc48c0?, 0xc000ade340})
	github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources/oauth_integration.go:313 +0xa74
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0x234a7a8?, {0x234a7a8?, 0xc000f4bb30?}, 0xd?, {0x1fc48c0?, 0xc000ade340?})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.31.0/helper/schema/resource.go:800 +0x163
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0007bb7a0, {0x234a7a8, 0xc000f4bb30}, 0xc000f365b0, 0xc000a28a80, {0x1fc48c0, 0xc000ade340})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.31.0/helper/schema/resource.go:919 +0x83a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc000a08558, {0x234a7a8?, 0xc000f4b980?}, 0xc000f41090)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.31.0/helper/schema/grpc_provider.go:1074 +0xdbc
github.com/hashicorp/terraform-plugin-mux/tf5to6server.v5tov6Server.ApplyResourceChange({{0x235fdc0?, 0xc000a08558?}}, {0x234a7a8, 0xc000f4b980}, 0x0?)
	github.com/hashicorp/terraform-plugin-mux@v0.13.0/tf5to6server/tf5to6server.go:37 +0x54
github.com/hashicorp/terraform-plugin-mux/tf6muxserver.(*muxServer).ApplyResourceChange(0x234a7e0?, {0x234a7a8?, 0xc000f4b680?}, 0xc000f41040)
	github.com/hashicorp/terraform-plugin-mux@v0.13.0/tf6muxserver/mux_server_ApplyResourceChange.go:36 +0x[193](https://github.com/cmctechnology/data-cdp-dw/actions/runs/7543033983/job/20533143587#step:30:194)
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).ApplyResourceChange(0xc000a023c0, {0x234a7a8?, 0xc000f4ae70?}, 0xc00099dce0)
	github.com/hashicorp/terraform-plugin-go@v0.20.0/tfprotov6/tf6server/server.go:857 +0x56b
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_ApplyResourceChange_Handler({0x1f76820?, 0xc000a023c0}, {0x234a7a8, 0xc000f4ae70}, 0xc000a28600, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.20.0/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:503 +0x169
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000dad20, {0x234a7a8, 0xc000f4ade0}, {0x235b0c0, 0xc00078e680}, 0xc000f527e0, 0xc00061d1d0, 0x330db58, 0x0)
	google.golang.org/grpc@v1.60.0/server.go:1372 +0xe03
google.golang.org/grpc.(*Server).handleStream(0xc0000dad20, {0x235b0c0, 0xc00078e680}, 0xc000f527e0)
	google.golang.org/grpc@v1.60.0/server.go:1783 +0xfec
google.golang.org/grpc.(*Server).serveStreams.func2.1()
	google.golang.org/grpc@v1.60.0/server.go:1016 +0x59
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 35
	google.golang.org/grpc@v1.60.0/server.go:1027 +0x115

Error: The terraform-provider-snowflake_v0.83.1 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

Error: Process completed with exit code 1.`

### Steps to Reproduce

terraform apply


### How much impact is this issue causing?

High

### Logs

_No response_

### Additional Information

I was initially authenticating user via private_key_path & private_key_passphrase. The last working version was 0.72.0.
Have recently figured out change in authentication, private key being deprecated so changed to password authentication. Below is what now provider looks like:

`provider "snowflake" {
  user    = var.snowflake_username
  account = local.snowflake_account
  # region                 = local.region
  role = var.snowflake_terraform_role
  # private_key_path       = var.SNOWFLAKE_PK_PATH
  # private_key_passphrase = var.SNOWFLAKE_PK_PASSPHRASE
  password = var.SNOWFLAKE_TERRAFORM_ACCESS_KEY
  # authenticator = "Snowflake"
}`

--- a/pkg/resources/database.go
+++ b/pkg/resources/database.go
@@ -263,16 +263,25 @@ func UpdateDatabase(d *schema.ResourceData, meta interface{}) error {
 	// If replication configuration changes, need to update accounts that have permission to replicate database
 	if d.HasChange("replication_configuration") {
 		oldConfig, newConfig := d.GetChange("replication_configuration")
-		newAccounts := newConfig.([]interface{})[0].(map[string]interface{})["accounts"].([]interface{})
-		newAccountIDs := make([]sdk.AccountIdentifier, len(newAccounts))
-		for i, account := range newAccounts {
-			newAccountIDs[i] = sdk.NewAccountIdentifierFromAccountLocator(account.(string))
+
+		newAccountIDs := make([]sdk.AccountIdentifier, 0)
+		ignoreEditionCheck := false
+		if len(newConfig.([]interface{})) != 0 {
+			newAccounts := newConfig.([]interface{})[0].(map[string]interface{})["accounts"].([]interface{})
+			for _, account := range newAccounts {
+				newAccountIDs = append(newAccountIDs, sdk.NewAccountIdentifierFromAccountLocator(account.(string)))
+			}
+			ignoreEditionCheck = newConfig.([]interface{})[0].(map[string]interface{})["ignore_edition_check"].(bool)
 		}
-		oldAccounts := oldConfig.([]interface{})[0].(map[string]interface{})["accounts"].([]interface{})
-		oldAccountIDs := make([]sdk.AccountIdentifier, len(oldAccounts))
-		for i, account := range oldAccounts {
-			oldAccountIDs[i] = sdk.NewAccountIdentifierFromAccountLocator(account.(string))
+
+		oldAccountIDs := make([]sdk.AccountIdentifier, 0)
+		if len(oldConfig.([]interface{})) != 0 {
+			oldAccounts := oldConfig.([]interface{})[0].(map[string]interface{})["accounts"].([]interface{})
+			for _, account := range oldAccounts {
+				oldAccountIDs = append(oldAccountIDs, sdk.NewAccountIdentifierFromAccountLocator(account.(string)))
+			}
 		}
+
 		accountsToRemove := make([]sdk.AccountIdentifier, 0)
 		accountsToAdd := make([]sdk.AccountIdentifier, 0)
 		// Find accounts to remove
@@ -294,7 +303,7 @@ func UpdateDatabase(d *schema.ResourceData, meta interface{}) error {
 					ToAccounts: accountsToAdd,
 				},
 			}
-			if ignoreEditionCheck := d.Get("ignore_edition_check").(bool); ignoreEditionCheck {
+			if ignoreEditionCheck {
 				opts.EnableReplication.IgnoreEditionCheck = sdk.Bool(ignoreEditionCheck)
 			}
 			err := client.Databases.AlterReplication(ctx, id, opts)

--- a/pkg/resources/database_acceptance_test.go
+++ b/pkg/resources/database_acceptance_test.go
@@ -49,6 +49,8 @@ func TestAcc_Database(t *testing.T) {
 	prefix := "tst-terraform" + strings.ToUpper(acctest.RandStringFromCharSet(10, acctest.CharSetAlpha))
 	prefix2 := "tst-terraform" + strings.ToUpper(acctest.RandStringFromCharSet(10, acctest.CharSetAlpha))
 
+	secondaryAccountName := getSecondaryAccount(t)
+
 	resource.ParallelTest(t, resource.TestCase{
 		Providers:    acc.TestAccProviders(),
 		PreCheck:     func() { acc.TestAccPreCheck(t) },
@@ -80,11 +82,25 @@ func TestAcc_Database(t *testing.T) {
 					resource.TestCheckResourceAttr("snowflake_database.db", "data_retention_time_in_days", "3"),
 				),
 			},
+			// ADD REPLICATION
+			// proves https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2369 error
+			{
+				Config: dbConfigWithReplication(prefix2, secondaryAccountName),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("snowflake_database.db", "name", prefix2),
+					resource.TestCheckResourceAttr("snowflake_database.db", "comment", "test comment 2"),
+					resource.TestCheckResourceAttr("snowflake_database.db", "data_retention_time_in_days", "3"),
+					resource.TestCheckResourceAttr("snowflake_database.db", "replication_configuration.#", "1"),
+					resource.TestCheckResourceAttr("snowflake_database.db", "replication_configuration.0.accounts.#", "1"),
+					resource.TestCheckResourceAttr("snowflake_database.db", "replication_configuration.0.accounts.0", secondaryAccountName),
+				),
+			},
 			// IMPORT
 			{
-				ResourceName:      "snowflake_database.db",
-				ImportState:       true,
-				ImportStateVerify: true,
+				ResourceName:            "snowflake_database.db",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"replication_configuration"},
 			},
 		},
 	})
@@ -155,6 +171,22 @@ resource "snowflake_database" "db" {
 	return fmt.Sprintf(s, prefix)
 }
 
+func dbConfigWithReplication(prefix string, secondaryAccountName string) string {
+	s := `
+resource "snowflake_database" "db" {
+	name = "%s"
+	comment = "test comment 2"
+	data_retention_time_in_days = 3
+	replication_configuration {
+		accounts = [
+			"%s"
+		]
+	}
+}
+`
+	return fmt.Sprintf(s, prefix, secondaryAccountName)
+}
+
 func dropDatabaseOutsideTerraform(t *testing.T, id string) {
 	t.Helper()
 
@@ -164,6 +196,23 @@ func dropDatabaseOutsideTerraform(t *testing.T, id string) {
 
 	err = client.Databases.Drop(ctx, sdk.NewAccountObjectIdentifier(id), &sdk.DropDatabaseOptions{})
 	require.NoError(t, err)
+}
+
+func getSecondaryAccount(t *testing.T) string {
+	t.Helper()
+
+	secondaryConfig, err := sdk.ProfileConfig("secondary_test_account")
+	require.NoError(t, err)
+
+	secondaryClient, err := sdk.NewClient(secondaryConfig)
+	require.NoError(t, err)
+
+	ctx := context.Background()
+
+	account, err := secondaryClient.ContextFunctions.CurrentAccount(ctx)
+	require.NoError(t, err)
+
+	return account
 }
 
 func testAccCheckDatabaseExistence(t *testing.T, id string, shouldExist bool) func(state *terraform.State) error {

--- a/pkg/resources/dynamic_table.go
+++ b/pkg/resources/dynamic_table.go
@@ -3,6 +3,7 @@ package resources
 import (
 	"context"
 	"database/sql"
+	"errors"
 	"log"
 	"strings"
 
@@ -190,8 +191,7 @@ func ReadDynamicTable(d *schema.ResourceData, meta interface{}) error {
 			return err
 		}
 	}
-	text := dynamicTable.Text
-	if strings.Contains(text, "OR REPLACE") {
+	if strings.Contains(dynamicTable.Text, "OR REPLACE") {
 		if err := d.Set("or_replace", true); err != nil {
 			return err
 		}
@@ -242,7 +242,43 @@ func ReadDynamicTable(d *schema.ResourceData, meta interface{}) error {
 	if err := d.Set("data_timestamp", dynamicTable.DataTimestamp.Format("2006-01-02T16:04:05.000 -0700")); err != nil {
 		return err
 	}
+
+	query, err := getQueryFromDDL(dynamicTable.Text)
+	if err != nil {
+		return err
+	}
+	if err := d.Set("query", query); err != nil {
+		return err
+	}
+
 	return nil
+}
+
+/*
+ * Previous implementation tried to match query part from the whole dynamic table DDL statement by just using `AS`.
+ * It was failing for table names containing `AS` (like `REASON`). It was also failing for other parts containing `AS`.
+ * We cannot simply match by ` AS ` because this can still be part of COMMENT or SELECT query itself.
+ * We have considered not setting the query at all but it was not ideal because of:
+ * - possible external changes to dynamic table (drop and recreate externally with different query);
+ * - import not 100% correct.
+ * We did not want to complicate the implementation too much by introducing parsers.
+ * One more thing worth mentioning is the whitespace that can be introduced by the user that is still returned by SHOW.
+ * For now, we just normalize the DDL before extraction of query.
+ *
+ * The outcome implementation matches by ` AS SELECT ` and checks the number of matches.
+ * If more matches are found, the error is returned to inform user about possible cause of error.
+ *
+ * Refer to issue https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2329.
+ */
+func getQueryFromDDL(text string) (string, error) {
+	normalizedDDL := normalizeQuery(text)
+	matchSubstring := " AS SELECT "
+	matches := strings.Count(strings.ToUpper(normalizedDDL), matchSubstring)
+	if matches != 1 {
+		return "", errors.New("too many matches found. There is no way of getting ONLY the 'query' used to create the dynamic table from Snowflake. We try to get it from the whole creation statement but there may be cases where it fails. Please submit the issue on Github (refer to #2329)")
+	}
+	idx := strings.Index(strings.ToUpper(normalizedDDL), " AS SELECT ")
+	return strings.TrimSpace(normalizedDDL[idx+4:]), nil
 }
 
 func parseTargetLag(v interface{}) sdk.TargetLag {

--- a/pkg/resources/dynamic_table.go
+++ b/pkg/resources/dynamic_table.go
@@ -200,11 +200,6 @@ func ReadDynamicTable(d *schema.ResourceData, meta interface{}) error {
 			return err
 		}
 	}
-	// trim up to " ..AS" and remove whitespace
-	query := strings.TrimSpace(text[strings.Index(text, "AS")+3:])
-	if err := d.Set("query", query); err != nil {
-		return err
-	}
 	if err := d.Set("cluster_by", dynamicTable.ClusterBy); err != nil {
 		return err
 	}

--- a/pkg/resources/dynamic_table_acceptance_test.go
+++ b/pkg/resources/dynamic_table_acceptance_test.go
@@ -284,6 +284,52 @@ func TestAcc_DynamicTable_issue2276(t *testing.T) {
 	})
 }
 
+// TestAcc_DynamicTable_issue2329 proves https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2329 issue.
+func TestAcc_DynamicTable_issue2329(t *testing.T) {
+	dynamicTableName := strings.ToUpper(acctest.RandStringFromCharSet(4, acctest.CharSetAlpha) + "AS" + acctest.RandStringFromCharSet(4, acctest.CharSetAlpha))
+	tableName := dynamicTableName + "_table"
+	query := fmt.Sprintf(`select "id" from "%v"."%v"."%v"`, acc.TestDatabaseName, acc.TestSchemaName, tableName)
+	m := func() map[string]config.Variable {
+		return map[string]config.Variable{
+			"name":       config.StringVariable(dynamicTableName),
+			"database":   config.StringVariable(acc.TestDatabaseName),
+			"schema":     config.StringVariable(acc.TestSchemaName),
+			"warehouse":  config.StringVariable(acc.TestWarehouseName),
+			"query":      config.StringVariable(query),
+			"comment":    config.StringVariable("Comment with AS on purpose"),
+			"table_name": config.StringVariable(tableName),
+		}
+	}
+
+	resource.Test(t, resource.TestCase{
+		ProtoV6ProviderFactories: acc.TestAccProtoV6ProviderFactories,
+		PreCheck:                 func() { acc.TestAccPreCheck(t) },
+		TerraformVersionChecks: []tfversion.TerraformVersionCheck{
+			tfversion.RequireAbove(tfversion.Version1_5_0),
+		},
+		CheckDestroy: testAccCheckDynamicTableDestroy,
+		Steps: []resource.TestStep{
+			{
+				ConfigDirectory: config.TestStepDirectory(),
+				ConfigVariables: m(),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "name", dynamicTableName),
+					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "query", query),
+				),
+			},
+			// No changes are expected
+			{
+				ConfigDirectory: acc.ConfigurationSameAsStepN(1),
+				ConfigVariables: m(),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "name", dynamicTableName),
+					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "query", query),
+				),
+			},
+		},
+	})
+}
+
 func testAccCheckDynamicTableDestroy(s *terraform.State) error {
 	db := acc.TestAccProvider.Meta().(*sql.DB)
 	client := sdk.NewClientFromDB(db)

--- a/pkg/resources/dynamic_table_acceptance_test.go
+++ b/pkg/resources/dynamic_table_acceptance_test.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"database/sql"
 	"fmt"
+	"regexp"
 	"strings"
 	"testing"
 
@@ -291,11 +292,12 @@ func TestAcc_DynamicTable_issue2329(t *testing.T) {
 	query := fmt.Sprintf(`select "id" from "%v"."%v"."%v"`, acc.TestDatabaseName, acc.TestSchemaName, tableName)
 	m := func() map[string]config.Variable {
 		return map[string]config.Variable{
-			"name":       config.StringVariable(dynamicTableName),
-			"database":   config.StringVariable(acc.TestDatabaseName),
-			"schema":     config.StringVariable(acc.TestSchemaName),
-			"warehouse":  config.StringVariable(acc.TestWarehouseName),
-			"query":      config.StringVariable(query),
+			"name":      config.StringVariable(dynamicTableName),
+			"database":  config.StringVariable(acc.TestDatabaseName),
+			"schema":    config.StringVariable(acc.TestSchemaName),
+			"warehouse": config.StringVariable(acc.TestWarehouseName),
+			// spaces added on purpose
+			"query":      config.StringVariable("  " + query),
 			"comment":    config.StringVariable("Comment with AS on purpose"),
 			"table_name": config.StringVariable(tableName),
 		}
@@ -325,6 +327,41 @@ func TestAcc_DynamicTable_issue2329(t *testing.T) {
 					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "name", dynamicTableName),
 					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "query", query),
 				),
+			},
+		},
+	})
+}
+
+// TestAcc_DynamicTable_issue2329_with_matching_comment proves https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2329 issue.
+func TestAcc_DynamicTable_issue2329_with_matching_comment(t *testing.T) {
+	dynamicTableName := strings.ToUpper(acctest.RandStringFromCharSet(4, acctest.CharSetAlpha) + "AS" + acctest.RandStringFromCharSet(4, acctest.CharSetAlpha))
+	tableName := dynamicTableName + "_table"
+	query := fmt.Sprintf(`select "id" from "%v"."%v"."%v"`, acc.TestDatabaseName, acc.TestSchemaName, tableName)
+	m := func() map[string]config.Variable {
+		return map[string]config.Variable{
+			"name":       config.StringVariable(dynamicTableName),
+			"database":   config.StringVariable(acc.TestDatabaseName),
+			"schema":     config.StringVariable(acc.TestSchemaName),
+			"warehouse":  config.StringVariable(acc.TestWarehouseName),
+			"query":      config.StringVariable(query),
+			"comment":    config.StringVariable("Comment with AS SELECT on purpose"),
+			"table_name": config.StringVariable(tableName),
+		}
+	}
+
+	resource.Test(t, resource.TestCase{
+		ProtoV6ProviderFactories: acc.TestAccProtoV6ProviderFactories,
+		PreCheck:                 func() { acc.TestAccPreCheck(t) },
+		TerraformVersionChecks: []tfversion.TerraformVersionCheck{
+			tfversion.RequireAbove(tfversion.Version1_5_0),
+		},
+		CheckDestroy: testAccCheckDynamicTableDestroy,
+		Steps: []resource.TestStep{
+			// If we match more than one time (in this case in comment) we raise an explanation error.
+			{
+				ConfigDirectory: acc.ConfigurationDirectory("TestAcc_DynamicTable_issue2329/1"),
+				ConfigVariables: m(),
+				ExpectError:     regexp.MustCompile(`too many matches found. There is no way of getting ONLY the 'query' used to create the dynamic table from Snowflake. We try to get it from the whole creation statement but there may be cases where it fails. Please submit the issue on Github \(refer to #2329\)`),
 			},
 		},
 	})

--- a/pkg/resources/oauth_integration.go
+++ b/pkg/resources/oauth_integration.go
@@ -310,7 +310,7 @@ func UpdateOAuthIntegration(d *schema.ResourceData, meta interface{}) error {
 
 	if d.HasChange("blocked_roles_list") {
 		runSetStatement = true
-		stmt.SetStringList(`BLOCKED_ROLES_LIST`, expandStringList(d.Get("blocked_roles_list").([]interface{})))
+		stmt.SetStringList(`BLOCKED_ROLES_LIST`, expandStringList(d.Get("blocked_roles_list").(*schema.Set).List()))
 	}
 
 	if d.HasChange("enabled") {

--- a/pkg/resources/oauth_integration_acceptance_test.go
+++ b/pkg/resources/oauth_integration_acceptance_test.go
@@ -21,7 +21,7 @@ func TestAcc_OAuthIntegration(t *testing.T) {
 		CheckDestroy: nil,
 		Steps: []resource.TestStep{
 			{
-				Config: oauthIntegrationConfig(name, oauthClient, clientType),
+				Config: oauthIntegrationConfig(name, oauthClient, clientType, "SYSADMIN"),
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("snowflake_oauth_integration.test", "name", name),
 					resource.TestCheckResourceAttr("snowflake_oauth_integration.test", "oauth_client", oauthClient),
@@ -33,6 +33,15 @@ func TestAcc_OAuthIntegration(t *testing.T) {
 				),
 			},
 			{
+				// role change proves https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2358 issue
+				Config: oauthIntegrationConfig(name, oauthClient, clientType, "USERADMIN"),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("snowflake_oauth_integration.test", "name", name),
+					resource.TestCheckResourceAttr("snowflake_oauth_integration.test", "blocked_roles_list.#", "1"),
+					resource.TestCheckResourceAttr("snowflake_oauth_integration.test", "blocked_roles_list.0", "USERADMIN"),
+				),
+			},
+			{
 				ResourceName:      "snowflake_oauth_integration.test",
 				ImportState:       true,
 				ImportStateVerify: true,
@@ -41,7 +50,7 @@ func TestAcc_OAuthIntegration(t *testing.T) {
 	})
 }
 
-func oauthIntegrationConfig(name, oauthClient, clientType string) string {
+func oauthIntegrationConfig(name, oauthClient, clientType string, blockedRole string) string {
 	return fmt.Sprintf(`
 	resource "snowflake_oauth_integration" "test" {
 		name                         = "%s"
@@ -51,9 +60,9 @@ func oauthIntegrationConfig(name, oauthClient, clientType string) string {
 		enabled                      = true
   		oauth_issue_refresh_tokens   = true
   		oauth_refresh_token_validity = 3600
-  		blocked_roles_list           = ["SYSADMIN"]
+  		blocked_roles_list           = ["%s"]
 	}
-	`, name, oauthClient, clientType)
+	`, name, oauthClient, clientType, blockedRole)
 }
 
 func TestAcc_OAuthIntegrationTableau(t *testing.T) {

--- a/pkg/resources/testdata/TestAcc_DynamicTable_issue2329/1/test.tf
+++ b/pkg/resources/testdata/TestAcc_DynamicTable_issue2329/1/test.tf
@@ -1,0 +1,27 @@
+resource "snowflake_table" "t" {
+  database        = var.database
+  schema          = var.schema
+  name            = var.table_name
+  change_tracking = true
+  column {
+    name = "id"
+    type = "NUMBER(38,0)"
+  }
+  column {
+    name = "data"
+    type = "VARCHAR(16)"
+  }
+}
+
+resource "snowflake_dynamic_table" "dt" {
+  depends_on = [snowflake_table.t]
+  name       = var.name
+  database   = var.database
+  schema     = var.schema
+  target_lag {
+    maximum_duration = "2 minutes"
+  }
+  warehouse = var.warehouse
+  query     = var.query
+  comment   = var.comment
+}

--- a/pkg/resources/testdata/TestAcc_DynamicTable_issue2329/1/variables.tf
+++ b/pkg/resources/testdata/TestAcc_DynamicTable_issue2329/1/variables.tf
@@ -1,0 +1,27 @@
+variable "name" {
+  type = string
+}
+
+variable "database" {
+  type = string
+}
+
+variable "schema" {
+  type = string
+}
+
+variable "warehouse" {
+  type = string
+}
+
+variable "query" {
+  type = string
+}
+
+variable "comment" {
+  type = string
+}
+
+variable "table_name" {
+  type = string
+}

--- a/pkg/sdk/client.go
+++ b/pkg/sdk/client.go
@@ -73,7 +73,7 @@ type Client struct {
 	Shares                   Shares
 	Stages                   Stages
 	StorageIntegrations      StorageIntegrations
-  Streamlits               Streamlits
+	Streamlits               Streamlits
 	Streams                  Streams
 	Tables                   Tables
 	Tags                     Tags

--- a/pkg/sdk/poc/main.go
+++ b/pkg/sdk/poc/main.go
@@ -36,7 +36,7 @@ var definitionMapping = map[string]*generator.Interface{
 	"materialized_views_def.go":        sdk.MaterializedViewsDef,
 	"api_integrations_def.go":          sdk.ApiIntegrationsDef,
 	"notification_integrations_def.go": sdk.NotificationIntegrationsDef,
-  "streamlits_def.go":                sdk.StreamlitsDef,
+	"streamlits_def.go":                sdk.StreamlitsDef,
 }
 
 func main() {

--- a/pkg/sdk/testint/streamlits_integration_test.go
+++ b/pkg/sdk/testint/streamlits_integration_test.go
@@ -55,7 +55,7 @@ func TestInt_Streamlits(t *testing.T) {
 	}
 
 	t.Run("create streamlit", func(t *testing.T) {
-		stage, cleanupStage := createStage(t, client, databaseTest, schemaTest, random.AlphaN(4))
+		stage, cleanupStage := createStage(t, client, sdk.NewSchemaObjectIdentifier(TestDatabaseName, TestSchemaName, random.AlphaN(4)))
 		t.Cleanup(cleanupStage)
 
 		comment := random.StringN(4)
@@ -70,7 +70,7 @@ func TestInt_Streamlits(t *testing.T) {
 	})
 
 	t.Run("alter streamlit: set", func(t *testing.T) {
-		stage, cleanupStage := createStage(t, client, databaseTest, schemaTest, random.AlphaN(4))
+		stage, cleanupStage := createStage(t, client, sdk.NewSchemaObjectIdentifier(TestDatabaseName, TestSchemaName, random.AlphaN(4)))
 		t.Cleanup(cleanupStage)
 		manifest := "manifest.yml"
 		e := createStreamlitHandle(t, stage, manifest)
@@ -84,7 +84,7 @@ func TestInt_Streamlits(t *testing.T) {
 	})
 
 	t.Run("alter function: rename", func(t *testing.T) {
-		stage, cleanupStage := createStage(t, client, databaseTest, schemaTest, random.AlphaN(4))
+		stage, cleanupStage := createStage(t, client, sdk.NewSchemaObjectIdentifier(TestDatabaseName, TestSchemaName, random.AlphaN(4)))
 		t.Cleanup(cleanupStage)
 		e := createStreamlitHandle(t, stage, "manifest.yml")
 
@@ -107,7 +107,7 @@ func TestInt_Streamlits(t *testing.T) {
 	})
 
 	t.Run("show streamlit: with like", func(t *testing.T) {
-		stage, cleanupStage := createStage(t, client, databaseTest, schemaTest, random.AlphaN(4))
+		stage, cleanupStage := createStage(t, client, sdk.NewSchemaObjectIdentifier(TestDatabaseName, TestSchemaName, random.AlphaN(4)))
 		t.Cleanup(cleanupStage)
 		e := createStreamlitHandle(t, stage, "manifest.yml")
 
@@ -118,7 +118,7 @@ func TestInt_Streamlits(t *testing.T) {
 	})
 
 	t.Run("show streamlit: terse with like", func(t *testing.T) {
-		stage, cleanupStage := createStage(t, client, databaseTest, schemaTest, random.AlphaN(4))
+		stage, cleanupStage := createStage(t, client, sdk.NewSchemaObjectIdentifier(TestDatabaseName, TestSchemaName, random.AlphaN(4)))
 		t.Cleanup(cleanupStage)
 		e := createStreamlitHandle(t, stage, "manifest.yml")
 
@@ -139,7 +139,7 @@ func TestInt_Streamlits(t *testing.T) {
 	})
 
 	t.Run("describe streamlit", func(t *testing.T) {
-		stage, cleanupStage := createStage(t, client, databaseTest, schemaTest, random.AlphaN(4))
+		stage, cleanupStage := createStage(t, client, sdk.NewSchemaObjectIdentifier(TestDatabaseName, TestSchemaName, random.AlphaN(4)))
 		t.Cleanup(cleanupStage)
 
 		mainFile := "manifest.yml"

